File: VB/GridView_PriorityRowCellStyle/CustomGridControl.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Text
Imports DevExpress.XtraGrid.Views.Grid
Imports DevExpress.XtraGrid.Views.Grid.Drawing
Imports DevExpress.XtraGrid.Registrator
Imports DevExpress.XtraGrid
Imports DevExpress.XtraGrid.Views.Base
Imports DevExpress.Utils
Imports DevExpress.XtraGrid.Views.Grid.ViewInfo
Imports DevExpress.XtraGrid.Columns
Imports System.ComponentModel

Namespace GridView_PriorityRowCellStyle
	Public Class CustomGridView
		Inherits GridView
		Private priorityStyleEventArgs As New PriorityRowCellStyleEventArgs(Nothing, New AppearanceObject(), False)
		Public Sub New()
			MyBase.New()
		End Sub
		Protected Friend Overridable Sub SetGridControlAccessMetod(ByVal newControl As GridControl)
			SetGridControl(newControl)
		End Sub

		Private Shared ReadOnly _priorityRowCellStyle As Object = New Object()

		Public Overrides Sub Assign(ByVal v As BaseView, ByVal copyEvents As Boolean)
			If v Is Nothing Then
				Return
			End If
			BeginUpdate()
			Try
				Dim gv As CustomGridView = TryCast(v, CustomGridView)
				If gv IsNot Nothing Then
					If copyEvents Then
						Events.AddHandler(_priorityRowCellStyle, gv.Events(_priorityRowCellStyle))
					End If
				End If
			Finally
				EndUpdate()
			End Try
		End Sub

		<Description("Enables the appearance settings of individual cells to be changed, or merged with Format Conditon"), Category("Appearance")> _
		Public Custom Event PriorityRowCellStyle As PriorityRowCellStyleEventHandler
			AddHandler(ByVal value As PriorityRowCellStyleEventHandler)
				Me.Events.AddHandler(_priorityRowCellStyle, value)
			End AddHandler
			RemoveHandler(ByVal value As PriorityRowCellStyleEventHandler)
				Me.Events.RemoveHandler(_priorityRowCellStyle, value)
			End RemoveHandler
			RaiseEvent(ByVal sender As Object, ByVal e As PriorityRowCellStyleEventArgs)
			End RaiseEvent
		End Event
		Protected Friend Overridable Function RaisePriorityRowCellStyle(ByVal cellInfo As GridCellInfo, ByVal appearance As AppearanceObject, ByVal overrydeFormatCondition As Boolean) As AppearanceObject
			priorityStyleEventArgs.OverrydeFormatCondition = False
			Dim handler As PriorityRowCellStyleEventHandler = CType(Events(_priorityRowCellStyle), PriorityRowCellStyleEventHandler)
			If handler Is Nothing Then
				Return appearance
			End If
			priorityStyleEventArgs.Appearance = TryCast(appearance.Clone(), AppearanceObject)
			priorityStyleEventArgs.SetCellInfo(cellInfo)
			handler(Me, priorityStyleEventArgs)
			Return priorityStyleEventArgs.Appearance
		End Function

		Protected Overrides ReadOnly Property ViewName() As String
			Get
				Return "CustomGridView"
			End Get
		End Property
		Protected Friend Overridable Function GetPriorityRowCellStyle(ByVal cellInfo As GridCellInfo, ByVal appearance As AppearanceObject) As AppearanceObject
			Return RaisePriorityRowCellStyle(cellInfo, appearance, False)
		End Function
		Protected Friend Overridable ReadOnly Property OverrideFormatCondition() As Boolean
			Get
				Return priorityStyleEventArgs.OverrydeFormatCondition
			End Get
		End Property
	End Class

	Public Delegate Sub PriorityRowCellStyleEventHandler(ByVal sender As Object, ByVal e As PriorityRowCellStyleEventArgs)
	Public Class PriorityRowCellStyleEventArgs
		Inherits EventArgs
		Private overrydeFormatCondition_Renamed As Boolean
		Private cellInfo_Renamed As GridCellInfo
		Private appearance_Renamed As AppearanceObject
		Public Sub New(ByVal cellInfo As GridCellInfo, ByVal appearance As AppearanceObject, ByVal overrydeFormatCondition As Boolean)
			MyBase.New()
			Me.overrydeFormatCondition_Renamed = overrydeFormatCondition
			Me.cellInfo_Renamed = cellInfo
			Me.appearance_Renamed = appearance
		End Sub
		Public Property OverrydeFormatCondition() As Boolean
			Get
				Return overrydeFormatCondition_Renamed
			End Get
			Set(ByVal value As Boolean)
				overrydeFormatCondition_Renamed = value
			End Set
		End Property
		Public ReadOnly Property CellInfo() As GridCellInfo
			Get
				Return cellInfo_Renamed
			End Get
		End Property
		Protected Friend Sub SetCellInfo(ByVal cellInfo As GridCellInfo)
			Me.cellInfo_Renamed = cellInfo
		End Sub
		Public Property Appearance() As AppearanceObject
			Get
				Return appearance_Renamed
			End Get
			Set(ByVal value As AppearanceObject)
				appearance_Renamed = value
			End Set
		End Property

	End Class
	Public Class CustomGridControl
		Inherits GridControl
		Public Sub New()
			MyBase.New()
		End Sub
		Protected Overrides Sub RegisterAvailableViewsCore(ByVal collection As InfoCollection)
			MyBase.RegisterAvailableViewsCore(collection)
			collection.Add(New CustomGridInfoRegistrator())
		End Sub
		Protected Overrides Function CreateDefaultView() As BaseView
			Return CreateView("CustomGridView")
		End Function

	End Class
	Public Class CustomGridInfoRegistrator
		Inherits GridInfoRegistrator
		Public Sub New()
			MyBase.New()
		End Sub
		Public Overrides ReadOnly Property ViewName() As String
			Get
				Return "CustomGridView"
			End Get
		End Property
		Public Overrides Function CreateView(ByVal grid As GridControl) As BaseView
			Dim view As New CustomGridView()
			view.SetGridControlAccessMetod(grid)
			Return view
		End Function


		Public Overrides Function CreateViewInfo(ByVal view As BaseView) As DevExpress.XtraGrid.Views.Base.ViewInfo.BaseViewInfo
			Return New CustomGridViewInfo(TryCast(view, GridView))
		End Function
	End Class
	Public Class CustomGridViewInfo
		Inherits GridViewInfo
        Public Sub New(ByVal gridView As GridView)
            MyBase.New(gridView)
        End Sub
        Protected Overrides Sub UpdateCellAppearanceCore(cell As GridCellInfo, Optional allowCache As Boolean = True, Optional allowCondition As Boolean = True, Optional cellCondition As AppearanceObjectEx = Nothing)
            MyBase.UpdateCellAppearanceCore(cell, allowCache, allowCondition, cellCondition)
            If cell.IsDataCell Then
                If (cell.State And GridRowCellState.FocusedCell) = 0 OrElse (Not View.IsEditing) Then
                    Dim condition As AppearanceObjectEx = Nothing
                    Dim priorityStyle As AppearanceObject = (CType(View, CustomGridView)).GetPriorityRowCellStyle(cell, cell.Appearance)
                    If priorityStyle IsNot cell.Appearance Then
                        If Not (CType(View, CustomGridView)).OverrideFormatCondition Then
                            condition = cell.RowInfo.ConditionInfo.GetCellAppearance(cell.Column)
                            If condition IsNot Nothing Then
                                priorityStyle = MergeAppearences(priorityStyle, condition, cell.Appearance)
                            End If
                        End If
                        cell.Appearance = priorityStyle
                    End If
                End If
            End If
        End Sub

        Protected Overridable Function MergeAppearences(ByVal target As AppearanceObject, ByVal source As AppearanceObject, ByVal current As AppearanceObject) As AppearanceObject
			If source.Options.UseBackColor Then
				If (Not source.BackColor.IsEmpty) AndAlso source.BackColor = current.BackColor Then
					target.BackColor = source.BackColor
				End If
				If (Not source.BackColor2.IsEmpty) AndAlso source.BackColor2 = current.BackColor2 Then
					target.BackColor2 = source.BackColor2
				End If
			End If
			If source.Options.UseBorderColor AndAlso source.BorderColor = current.BorderColor Then
				target.BorderColor = source.BorderColor
			End If
			If source.Options.UseFont AndAlso source.Font Is current.Font Then
				target.Font = source.Font
			End If
			If source.Options.UseForeColor AndAlso source.ForeColor = current.ForeColor Then
				target.ForeColor = source.ForeColor
			End If
            If source.Options.UseImage And source.Image Is current.Image Then
                target.Image = source.Image
            End If
			If source.Options.UseTextOptions AndAlso source.TextOptions Is current.TextOptions Then
				target.TextOptions.Assign(source.TextOptions)
			End If
			If source.GradientMode = current.GradientMode Then
				target.GradientMode = source.GradientMode
			End If
			Return target
		End Function
	End Class
End Namespace

File: CS/GridView_PriorityRowCellStyle/CustomGridControl.cs
using System;
using System.Collections.Generic;
using System.Text;
using DevExpress.XtraGrid.Views.Grid;
using DevExpress.XtraGrid.Views.Grid.Drawing;
using DevExpress.XtraGrid.Registrator;
using DevExpress.XtraGrid;
using DevExpress.XtraGrid.Views.Base;
using DevExpress.Utils;
using DevExpress.XtraGrid.Views.Grid.ViewInfo;
using DevExpress.XtraGrid.Columns;
using System.ComponentModel;

namespace GridView_PriorityRowCellStyle
{
    public class CustomGridView : GridView
    {
        PriorityRowCellStyleEventArgs priorityStyleEventArgs = new PriorityRowCellStyleEventArgs(null, new AppearanceObject(), false);
        public CustomGridView() : base() { }
        protected internal virtual void SetGridControlAccessMetod(GridControl newControl) { SetGridControl(newControl); }

        private static readonly object _priorityRowCellStyle = new object();

        public override void Assign(BaseView v, bool copyEvents)
        {
            if (v == null) return;
            BeginUpdate();
            try
            {
                CustomGridView gv = v as CustomGridView;
                if (gv != null)
                    if (copyEvents)
                        Events.AddHandler(_priorityRowCellStyle, gv.Events[_priorityRowCellStyle]);
            }
            finally { EndUpdate(); }
        }

        [Description("Enables the appearance settings of individual cells to be changed, or merged with Format Conditon"), Category("Appearance")]
        public event PriorityRowCellStyleEventHandler PriorityRowCellStyle
        {
            add { this.Events.AddHandler(_priorityRowCellStyle, value); }
            remove { this.Events.RemoveHandler(_priorityRowCellStyle, value); }
        }
        protected internal virtual AppearanceObject RaisePriorityRowCellStyle(GridCellInfo cellInfo, AppearanceObject appearance, bool overrydeFormatCondition)
        {
            priorityStyleEventArgs.OverrydeFormatCondition = false;
            PriorityRowCellStyleEventHandler handler = (PriorityRowCellStyleEventHandler)Events[_priorityRowCellStyle];
            if (handler == null) return appearance;
            priorityStyleEventArgs.Appearance = appearance.Clone() as AppearanceObject;
            priorityStyleEventArgs.SetCellInfo(cellInfo);
            handler(this, priorityStyleEventArgs);
            return priorityStyleEventArgs.Appearance;
        }

        protected override string ViewName { get { return "CustomGridView"; } }
        protected internal virtual AppearanceObject GetPriorityRowCellStyle(GridCellInfo cellInfo, AppearanceObject appearance)
        {
            return RaisePriorityRowCellStyle(cellInfo , appearance, false);
        }
        protected internal virtual bool OverrideFormatCondition { get { return priorityStyleEventArgs.OverrydeFormatCondition; } }
    }

    public delegate void PriorityRowCellStyleEventHandler(object sender, PriorityRowCellStyleEventArgs e);
    public class PriorityRowCellStyleEventArgs : EventArgs
    {
        bool overrydeFormatCondition;
        GridCellInfo cellInfo;
        AppearanceObject appearance;
        public PriorityRowCellStyleEventArgs(GridCellInfo cellInfo, AppearanceObject appearance, bool overrydeFormatCondition)
            : base()
        {
            this.overrydeFormatCondition = overrydeFormatCondition;
            this.cellInfo = cellInfo;
            this.appearance = appearance;
        }
        public bool OverrydeFormatCondition
        {
            get { return overrydeFormatCondition; }
            set { overrydeFormatCondition = value; }
        }
        public GridCellInfo CellInfo { get { return cellInfo; } }
        protected internal void SetCellInfo(GridCellInfo cellInfo) { this.cellInfo = cellInfo; }
        public AppearanceObject Appearance
        {
            get { return appearance; }
            set { appearance = value; }
        }

    }
    public class CustomGridControl : GridControl
    {
        public CustomGridControl() : base() { }
        protected override void RegisterAvailableViewsCore(InfoCollection collection)
        {
            base.RegisterAvailableViewsCore(collection);
            collection.Add(new CustomGridInfoRegistrator());
        }
        protected override BaseView CreateDefaultView() { return CreateView("CustomGridView"); }

    }
    public class CustomGridInfoRegistrator : GridInfoRegistrator
    {
        public CustomGridInfoRegistrator() : base() { }
        public override string ViewName { get { return "CustomGridView"; } }
        public override BaseView CreateView(GridControl grid)
        {
            CustomGridView view = new CustomGridView();
            view.SetGridControlAccessMetod(grid);
            return view;
        }


        public override DevExpress.XtraGrid.Views.Base.ViewInfo.BaseViewInfo CreateViewInfo(BaseView view) { return new CustomGridViewInfo(view as GridView); }
    }
    public class CustomGridViewInfo : GridViewInfo
    {
        public CustomGridViewInfo(GridView gridView) : base(gridView) { }
        protected override void UpdateCellAppearanceCore(GridCellInfo cell, bool allowCache = true, bool allowCondition = true, AppearanceObjectEx cellCondition = null)
        {
            base.UpdateCellAppearanceCore(cell, allowCache, allowCondition, cellCondition);
            if (cell.IsDataCell)
                if ((cell.State & GridRowCellState.FocusedCell) == 0 || !View.IsEditing)
                {
                    AppearanceObjectEx condition = null;
                    AppearanceObject priorityStyle = ((CustomGridView)View).GetPriorityRowCellStyle(cell, cell.Appearance);
                    if (priorityStyle != cell.Appearance)
                    {
                        if (!((CustomGridView)View).OverrideFormatCondition)
                        {
                            condition = cell.RowInfo.ConditionInfo.GetCellAppearance(cell.Column);
                            if (condition != null)
                                priorityStyle = MergeAppearences(priorityStyle, condition, cell.Appearance);
                        }
                        cell.Appearance = priorityStyle;
                    }
                }
        }

        protected virtual AppearanceObject MergeAppearences(AppearanceObject target, AppearanceObject source, AppearanceObject current)
        {
            if (source.Options.UseBackColor)
            {
                if (!source.BackColor.IsEmpty && source.BackColor == current.BackColor)
                    target.BackColor = source.BackColor;
                if (!source.BackColor2.IsEmpty && source.BackColor2 == current.BackColor2)
                    target.BackColor2 = source.BackColor2;
            }
            if (source.Options.UseBorderColor && source.BorderColor == current.BorderColor)
                target.BorderColor = source.BorderColor;
            if (source.Options.UseFont && source.Font == current.Font)
                target.Font = source.Font;
            if (source.Options.UseForeColor && source.ForeColor == current.ForeColor)
                target.ForeColor = source.ForeColor;
            if (source.Options.UseImage && source.Image == current.Image)
                target.Image = source.Image;
            if (source.Options.UseTextOptions && source.TextOptions == current.TextOptions)
                target.TextOptions.Assign(source.TextOptions);
            if (source.GradientMode == current.GradientMode)
                target.GradientMode = source.GradientMode;
            return target;
        }
    }
}
